walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,113 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
 
+/** A class to determine appropriate meeting times */
 public final class FindMeetingQuery {
+
+  /** Returns a list of time ranges such that a meeting with the specified constraints can be held
+  within each time range while still respecting the event calendar */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    long meetingDuration = request.getDuration(); 
+    Collection<String> mandatoryMeetingAttendees = request.getAttendees();
+    Collection<String> optionalMeetingAttendees = request.getOptionalAttendees();
+
+    // Return no availability if the requested meeting's duration is longer than one day
+    if (meetingDuration > (24 * 60)) {
+        return Arrays.asList();
+    }
+
+    List<TimeRange> mandatoryRanges = determineAvailability(events, mandatoryMeetingAttendees, meetingDuration);
+    List<TimeRange> optionalRanges = determineAvailability(events, optionalMeetingAttendees, meetingDuration);
+    
+    // Return availability based upon the quantity of mandatory verus optional attendees 
+    if (mandatoryMeetingAttendees.isEmpty() && optionalMeetingAttendees.isEmpty()) {
```

Comment:
Here the else if/else clauses are not adding extra information, as the function immediately returns when any statement is `true`. You could reduce some complexity and mental gymnastics by removing the `else` clauses

```java
if (thing1) {
  return "1";
} if (thing2) {
  return "2";
} if (thing3) {
  return "3";
}
// "else"
do.theThing();
return thing;
```

https://drive.google.com/file/d/0B-KFDiTrQe9lLU9uVDZzZWNmRVU/view

Reply:
adjusted